exchanges/bitstamp/bitstamp_websocket.go

```diff
@@ -232,66 +238,40 @@ func (b *Bitstamp) handleWSOrder(wsResp *websocketResponse, msg []byte) error {
-		return nil, err
-	}
-	var subscriptions subscription.List
-	for i := range enabledCurrencies {
-		p, err := b.FormatExchangeCurrency(enabledCurrencies[i], asset.Spot)
-		if err != nil {
-			return nil, err
-		}
-		for j := range defaultSubChannels {
-			subscriptions = append(subscriptions, &subscription.Subscription{
-				Channel: defaultSubChannels[j] + "_" + p.String(),
-				Asset:   asset.Spot,
-				Pairs:   currency.Pairs{p},
-			})
-		}
-		if b.Websocket.CanUseAuthenticatedEndpoints() {
-			for j := range defaultAuthSubChannels {
-				subscriptions = append(subscriptions, &subscription.Subscription{
-					Channel: defaultAuthSubChannels[j] + "_" + p.String(),
-					Asset:   asset.Spot,
-					Pairs:   currency.Pairs{p},
-					Params: map[string]interface{}{
-						"auth": struct{}{},
-					},
-				})
-			}
-		}
-	}
-	return subscriptions, nil
+func (b *Bitstamp) generateSubscriptions() (subscription.List, error) {
+	return b.Features.Subscriptions.ExpandTemplates(b)
 }
 
-// Subscribe sends a websocket message to receive data from the channel
-func (b *Bitstamp) Subscribe(channelsToSubscribe subscription.List) error {
-	var errs error
-	var auth *WebsocketAuthResponse
+// GetSubscriptionTemplate returns a subscription channel template
+func (b *Bitstamp) GetSubscriptionTemplate(_ *subscription.Subscription) (*template.Template, error) {
+	return template.New("master.tmpl").Funcs(template.FuncMap{"channelName": channelName}).Parse(subTplText)
+}
 
-	for i := range channelsToSubscribe {
-		if _, ok := channelsToSubscribe[i].Params["auth"]; ok {
-			var err error
-			auth, err = b.FetchWSAuth(context.TODO())
-			if err != nil {
-				errs = common.AppendError(errs, err)
-			}
-			break
-		}
+// Subscribe sends a websocket message to receive data from a list of channels
+func (b *Bitstamp) Subscribe(subs subscription.List) error {
+	var errs error
+	var creds *WebsocketAuthResponse
+	if authed := subs.Private(); len(authed) > 0 {
+		creds, errs = b.FetchWSAuth(context.TODO())
 	}
+	return common.AppendError(errs, b.ParallelChanOp(subs, func(s subscription.List) error { return b.subscribe(s, creds) }, 1))
+}
 
-	for _, s := range channelsToSubscribe {
+func (b *Bitstamp) subscribe(subs subscription.List, creds *WebsocketAuthResponse) error {
+	subs, errs := subs.ExpandTemplates(b)
+	for _, s := range subs {
 		req := websocketEventRequest{
 			Event: "bts:subscribe",
 			Data: websocketData{
-				Channel: s.Channel,
+				Channel: s.QualifiedChannel,
 			},
 		}
-		if _, ok := s.Params["auth"]; ok && auth != nil {
-			req.Data.Channel = "private-" + req.Data.Channel + "-" + strconv.Itoa(int(auth.UserID))
-			req.Data.Auth = auth.Token
+		if s.Authenticated {
+			if creds == nil {
+				return request.ErrAuthRequestFailed
+			}
+			req.Data.Channel = "private-" + req.Data.Channel + "-" + strconv.Itoa(int(creds.UserID))
+			req.Data.Auth = creds.Token
 		}
 		err := b.Websocket.Conn.SendJSONMessage(context.TODO(), request.Unset, req)
```

Comment:
Because they're lying:
```
> {"event":"bts:subscribe","data":{"channel":"live_trades_mouthfullmonkeygenetics_ethbtc"}}
< {"event":"bts:subscription_succeeded","channel":"live_trades_mouthfullmonkeygenetics_ethbtc","data":{}}
``` 
So that caused me to not bother delaying processing.
However, I think you're right. Let's DoTheRightThang and if they ever employ anyone that's actually competent maybe they'll fix it.

Reply:
Given BadCat, I wouldn' be surprised if that was a real trading pair you found

Reply:
Fixed https://github.com/gbjk/gocryptotrader/commit/033524ae3